### PR TITLE
Makes admin API friendly

### LINKED ISSUE #1116: Make administration endpoints API friendly
app.get('/admin', requireAdmin, views.admin.status)
  app.get('/admin/graphs', requireAdmin, views.admin.graphs)
  app.get('/admin/log', requireAdmin, views.admin.log)
  app.get('/admin/plugins', requireAdmin, views.admin.plugins)
  app.get('/admin/remotes', requireAdmin, views.admin.remotes)
  app.get('/admin/users', requireAdmin, views.admin.users)
  app.get('/admin/newUser', requireAdmin, views.admin.newUser)
  app.get('/admin/update', requireAdmin, views.admin.update)
  app.get('/admin/theme', requireAdmin, views.admin.theme)
  app.post('/admin/theme', requireAdmin, uploadToMemory.single('logo'), views.admin.theme)
  app.post('/admin/explorerUpdateIndex', requireAdmin, actions.admin.explorerUpdateIndex)
  app.get('/admin/explorer', requireAdmin, views.admin.explorer)
  app.post('/admin/explorer', requireAdmin, bodyParser.urlencoded({ extended: true }), views.admin.explorer)
  app.get('/admin/backup', requireAdmin, views.admin.backup)
  app.get('/admin/registries', requireAdmin, views.admin.registries)
  app.post('/admin/backup', requireAdmin, bodyParser.urlencoded({ extended: true }), views.admin.backup)
  app.post('/admin/users', requireAdmin, views.admin.users)
  app.post('/admin/backup/restore/:prefix', requireAdmin, bodyParser.urlencoded({ extended: true }), views.admin.backupRestore)
  app.post('/admin/newUser', requireAdmin, bodyParser.urlencoded({ extended: true }), views.admin.newUser)
  app.post('/admin/updateUser', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.updateUser)
  app.post('/admin/deleteUser', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.deleteUser)
  app.post('/admin/deleteRemote', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.deleteRemote)
  app.post('/admin/saveRemote', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.saveRemote)
  app.post('/admin/saveRegistry', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.saveRegistry)
  app.post('/admin/deleteRegistry', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.deleteRegistry)
  app.post('/admin/savePlugin', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.savePlugin)
  app.post('/admin/deletePlugin', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.deletePlugin)
  app.get('/admin/mail', requireAdmin, views.admin.mail)
  app.post('/admin/mail', requireAdmin, bodyParser.urlencoded({ extended: true }), views.admin.mail)
  app.post('/admin/setAdministratorEmail', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.setAdministratorEmail)
  app.post('/admin/explorerLog', requireAdmin, actions.admin.explorerLog)
  app.post('/updateWebOfRegistries', bodyParser.json(), api.updateWebOfRegistries)
  app.post('/admin/federate', requireAdmin, bodyParser.urlencoded({ extended: true }), actions.admin.federate)
  app.post('/admin/retrieveFromWebOfRegistries', requireAdmin, actions.admin.retrieve)
  app.get('/admin/sparql', requireAdmin, sparqlAdmin)
  app.post('/admin/sparql', requireAdmin, bodyParser.urlencoded({ extended: true }), sparqlAdmin)

### PR DESCRIPTION
**Pull request by cjmyers**
Closes #1116

### CODE REVIEW

File: lib/views/admin/explorer.js [hunk 1]
```diff
@@ -110,8 +122,16 @@ function post (req, res) {
...
-    form(req, res)
+    if (!req.accepts('text/html')) {
+      return res.status(200).header('content-type', 'text/plain').send('SBOLExplorer configuration updated successfully')
+    } else {
+      form(req, res)
+    }
   }).catch(() => {
-    form(req, res)
+    if (!req.accepts('text/html')) {
+      return res.status(404).header('content-type', 'text/plain').send('Error updating SBOLExplorer configuration')
```
Comment (3ach) [inline lib/views/admin/explorer.js:132]: 404?

File: package.json [hunk 2]
```diff
@@ -78,7 +78,7 @@
     "validator": "^10.9.0",
     "value-event": "^5.1.1",
     "virtual-dom": "^2.1.1",
-    "visbol": "^1.4.24",
+    "visbol": "^1.4.28",
```
Comment (3ach) [inline package.json:81]: ?

File: lib/views/admin/newUser.js [hunk 3]
```diff
@@ -16,15 +16,18 @@ module.exports = function (req, res) {
...
-    config: config.get(),
-    section: 'admin',
-    adminSection: 'theme',
-    user: req.user,
-    form: locals.form || {}
-  }, locals)
-
-  res.send(pug.renderFile('templates/views/admin/newUser.jade', locals))
+  if (!req.accepts('text/html')) {
+    return res.status(200).header('content-type', 'application/json').send(JSON.stringify(locals))
```
Comment (3ach) [inline lib/views/admin/newUser.js:20]: `locals` can contain the entire config, which contains values that we don't want to leak (like the synbiohub secret used to generate share links)

File: lib/views/admin/users.js [hunk 4]
```diff
@@ -18,15 +18,20 @@ module.exports = function (req, res) {
...
-      config: config.get(),
-      section: 'admin',
-      adminSection: 'users',
-      user: req.user,
       users: users,
-      canSendEmail: config.get('mail').sendgridApiKey != ''
+      canSendEmail: config.get('mail').sendgridApiKey !== ''
+    }
+    if (!req.accepts('text/html')) {
+      return res.status(200).header('content-type', 'application/json').send(JSON.stringify(locals))
```
Comment (3ach) [inline lib/views/admin/users.js:25]: Don't just send `locals`

File: lib/views/admin/theme.js [hunk 5]
```diff
@@ -29,18 +30,23 @@ function form (req, res) {
...
-    section: 'admin',
-    adminSection: 'theme',
-    user: req.user,
     currentTheme: currentTheme,
     themeParameters: themeParameters
   }
-
-  res.send(pug.renderFile('templates/views/admin/theme.jade', locals))
+  if (!req.accepts('text/html')) {
+    return res.status(200).header('content-type', 'application/json').send(JSON.stringify(locals))
```
Comment (3ach) [inline lib/views/admin/theme.js:40]: `locals` has secure values

File: lib/views/admin/status.js [hunk 6]
```diff
@@ -3,19 +3,24 @@ var pug = require('pug')
...
-    user: req.user,
     nodeVersion: process.version,
     architecture: os.arch(),
     platform: os.type(),
-    osRelease: os.release(),
-    config: config.get()
+    osRelease: os.release()
+  }
+  if (!req.accepts('text/html')) {
+    return res.status(200).header('content-type', 'application/json').send(JSON.stringify(locals))
```
Comment (3ach) [inline lib/views/admin/status.js:17]: `locals` has config

File: lib/views/admin/remotes.js [hunk 7]
```diff
@@ -1,27 +1,28 @@
...
-    section: 'admin',
-    adminSection: 'remotes',
-    user: req.user,
     remotes: remotes,
     remoteTypes: ['ice', 'benchling']
   }
-
-  res.send(pug.renderFile('templates/views/admin/remotes.jade', locals))
+  if (!req.accepts('text/html')) {
+    return res.status(200).header('content-type', 'application/json').send(JSON.stringify(locals))
```
Comment (3ach) [inline lib/views/admin/remotes.js:18]: locals issue

File: lib/views/admin/registries.js [hunk 8]
```diff
@@ -100,16 +103,21 @@ function unregistered (req, res, errors) {
...
     registries: registries,
     registered: false,
     wor: config.get('webOfRegistriesUrl'),
     worId: config.get('webOfRegistriesId'),
     errors: errors
   }
-
-  res.send(pug.renderFile('templates/views/admin/registries.jade', locals))
+  if (!req.accepts('text/html')) {
+    return res.status(200).header('content-type', 'application/json').send(JSON.stringify(locals))
```
Comment (3ach) [inline lib/views/admin/registries.js:113]: locals issue

File: lib/views/admin/registries.js [hunk 9]
```diff
@@ -80,8 +74,17 @@ module.exports = function (req, res) {
           approved: body['approved'],
           updateWorking: body['updateWorking']
         }
-
-        res.send(pug.renderFile('templates/views/admin/registries.jade', locals))
+        if (!req.accepts('text/html')) {
+          return res.status(200).header('content-type', 'application/json').send(JSON.stringify(locals))
```
Comment (3ach) [inline lib/views/admin/registries.js:78]: locals issue

**Comment by 3ach** [comment]
Overall these changes are very useful!!

File: lib/actions/admin/deletePlugin.js [hunk 10]
```diff
@@ -1,16 +1,30 @@
...
 
-    plugins[category].splice(id, 1)
+  if (!category || (category !== 'rendering' && category !== 'download' && category !== 'submit')) {
+    return res.status(400).header('content-type', 'text/plain').send('Must provide a category with value of rendering, submit, or download')
+  }
+  if (!id) {
+    return res.status(400).header('content-type', 'text/plain').send('Must provide a valid plugin id')
+  }
+  let index = parseInt(id, 10) - 1
+  if (!plugins || !plugins[category] || !plugins[category][index]) {
```
[the remark this answers is not in the document]
Comment (cjmyers) [reply lib/actions/admin/deletePlugin.js:17]: Agreed

File: lib/actions/admin/deleteRegistry.js [hunk 11]
```diff
@@ -5,8 +5,17 @@ module.exports = function (req, res) {
   const uri = req.body.uri
   var registries = config.get('webOfRegistries')
 
+  if (!uri || !registries || !registries[uri]) {
```
[the remark this answers is not in the document]
Comment (cjmyers) [reply lib/actions/admin/deleteRegistry.js:8]: Agreed

File: lib/actions/admin/deleteRemote.js [hunk 12]
```diff
@@ -5,8 +5,17 @@ module.exports = function (req, res) {
   const remoteId = req.body.id
   var remotes = config.get('remotes')
 
+  if (!remoteId || !remotes || !remotes[remoteId]) {
```
[the remark this answers is not in the document]
Comment (cjmyers) [reply lib/actions/admin/deleteRemote.js:8]: Agreed

File: lib/actions/admin/deleteUser.js [hunk 13]
```diff
@@ -3,10 +3,14 @@ const db = require('../../db')
 
 module.exports = function (req, res) {
   db.model.User.findById(req.body.id).then((user) => {
-    user.destroy()
+    if (user) {
+      user.destroy()
+    } else {
+      return res.status(400).send('Must provide a valid user id')
```
[the remark this answers is not in the document]
Comment (cjmyers) [reply lib/actions/admin/deleteUser.js:9]: Having an issue with this one see #1212

[the remark this answers is not in the document]
Comment (cjmyers) [reply lib/actions/admin/deleteUser.js:9]: Will remove the return

File: lib/actions/admin/deleteUser.js [hunk 14]
```diff
@@ -3,10 +3,14 @@ const db = require('../../db')
...
+    }
   }).then(() => {
-    res.status(200).send('saved')
+    res.status(200).send('user (' + req.body.id + ') deleted')
   }).catch((err) => {
-    res.status(500).send(err.stack)
+    return res.status(500).send(err.stack)
```
[the remark this answers is not in the document]
Comment (cjmyers) [reply lib/actions/admin/deleteUser.js:14]: Ditto.  Not sure why I'm getting unhandled promise rejection.

File: lib/actions/admin/savePlugin.js [hunk 15]
```diff
@@ -8,12 +8,30 @@ module.exports = function (req, res) {
...
+    return res.status(400).header('content-type', 'text/plain').send('Must provide a category with value of rendering, submit, or download')
+  }
+  if (!name) {
+    return res.status(400).header('content-type', 'text/plain').send('Must provide a valid plugin name')
+  }
+  if (!id) {
+    return res.status(400).header('content-type', 'text/plain').send('Must provide a valid plugin id')
+  } else if (id !== 'New') {
+    let index = parseInt(id, 10) - 1
+    if (!plugins || !plugins[category] || !plugins[category][index]) {
```
[the remark this answers is not in the document]
Comment (cjmyers) [reply lib/actions/admin/savePlugin.js:23]: Fixed

File: lib/views/admin/explorer.js (hunk 1 above)
Comment (cjmyers) [reply lib/views/admin/explorer.js:132]: Changed to 500

File: lib/views/admin/newUser.js (hunk 3 above)
Comment (cjmyers) [reply lib/views/admin/newUser.js:20]: Good catch.  Will just sent registerAlert.  BTW: must be admin to use these endpoints, but still see your point.

File: lib/views/admin/registries.js (hunk 9 above)
Comment (cjmyers) [reply lib/views/admin/registries.js:78]: This locals does not have config.

File: lib/views/admin/registries.js (hunk 8 above)
Comment (cjmyers) [reply lib/views/admin/registries.js:113]: This locals does not have config

File: lib/views/admin/remotes.js (hunk 7 above)
Comment (cjmyers) [reply lib/views/admin/remotes.js:18]: This locals does not have config

File: lib/views/admin/theme.js (hunk 5 above)
Comment (cjmyers) [reply lib/views/admin/theme.js:40]: This locals does not have config

File: lib/views/admin/users.js (hunk 4 above)
Comment (cjmyers) [reply lib/views/admin/users.js:25]: This locals does not have config

File: lib/views/admin/status.js (hunk 6 above)
Comment (cjmyers) [reply lib/views/admin/status.js:17]: Yep, this is the full config.  I've reduced it to just the config elements that appear in status plus a few more which are not currently on that page, but nothing private.

File: package.json (hunk 2 above)
Comment (cjmyers) [reply package.json:81]: Latest version of VisBol

**Comment by cjmyers** [comment]
@3ach Addressed your comments I think.  Not all locals leaks config, but I've fixed the ones that did.

File: lib/actions/admin/deleteUser.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (3ach) [reply lib/actions/admin/deleteUser.js]: I don't understand what you mean.

**Comment by cjmyers** [comment]
The issue with the promise rejection is described in #1212.  I will merge these changes, and you can reproduce the issue.  We can try to sort it when we meet tomorrow.  Indeed, I need yet another refresher on promises.

**Pull request merged**